Script to alter the til tracking numpy arrays if an error has occurred

In [1]:
# Standard packages
import tempfile
import warnings
import urllib
import shutil
import os
# Less standard, but still pip- or conda-installable
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
#import rasterio
import re
import rtree
import shapely
import pickle
import data_eng.az_proc as ap
#from cartopy import crs
import collections

In [2]:
import PIL
print('PIL',PIL.__version__)
from PIL import Image
Image.MAX_IMAGE_PIXELS = 999999999999

PIL 8.3.1


Specify the parameters to download images

In [3]:
parent_directory = "C:\chips_allocation"
annotation_directory = "Check"
number_of_tiles = 60
tiles_remaining = "tile_name_tile_url_remaining_expanded.npy"
tiles_labeled = "tile_name_tile_url_labeled.npy"

Load the annotator class

In [6]:
dist = ap.annotator(annotation_directory) #create the processing class
dist.state_dcc_directory(parent_directory)
dist.number_of_tiles(number_of_tiles)
dist.get_tile_urls(tiles_remaining)
dist.make_subdirectories()
dist.download_images()
#dist.tile_rename()

Unlabeled Tiles (545, 2)
0
Bypassing download of already-downloaded file m_3210310_ne_13_060_20180512_20190427.tif
1
Bypassing download of already-downloaded file m_4307023_sw_19_060_20180927.tif
2
Bypassing download of already-downloaded file m_3210312_se_13_060_20180506_20190427.tif
3
Bypassing download of already-downloaded file m_4011110_sw_12_060_20180909.tif
4
Bypassing download of already-downloaded file m_3907652_ne_18_060_20180828.tif
5
Bypassing download of already-downloaded file m_3109637_ne_14_060_20181102_20190228.tif
6
Bypassing download of already-downloaded file m_3110110_ne_14_060_20181113_20190504.tif
7
Bypassing download of already-downloaded file m_3808542_se_16_060_20180630.tif
8
Bypassing download of already-downloaded file m_4207148_ne_19_060_20181016.tif
9
Bypassing download of already-downloaded file m_3209933_se_14_060_20181028.tif
10
Bypassing download of already-downloaded file m_3411760_nw_11_060_20180822_20190209.tif
11
Bypassing download of already-downl

100% (565313695 of 565313695) |##########| Elapsed Time: 0:00:17 Time:  0:00:17


...done, 565313695 bytes.
13


100% (501721465 of 501721465) |##########| Elapsed Time: 0:00:13 Time:  0:00:13


...done, 501721465 bytes.
14


100% (538225862 of 538225862) |##########| Elapsed Time: 0:00:16 Time:  0:00:16


...done, 538225862 bytes.
15


100% (588940418 of 588940418) |##########| Elapsed Time: 0:00:18 Time:  0:00:18


...done, 588940418 bytes.
16


100% (593128142 of 593128142) |##########| Elapsed Time: 0:00:15 Time:  0:00:15
  0% (0 of 546272463) |                  | Elapsed Time: 0:00:00 ETA:  --:--:--

...done, 593128142 bytes.
17


100% (546272463 of 546272463) |##########| Elapsed Time: 0:00:12 Time:  0:00:12


...done, 546272463 bytes.
18


100% (646246675 of 646246675) |##########| Elapsed Time: 0:00:10 Time:  0:00:10


...done, 646246675 bytes.
19


100% (585782238 of 585782238) |##########| Elapsed Time: 0:00:21 Time:  0:00:21


...done, 585782238 bytes.
20


100% (493594469 of 493594469) |##########| Elapsed Time: 0:00:12 Time:  0:00:12


...done, 493594469 bytes.
21


100% (499414571 of 499414571) |##########| Elapsed Time: 0:00:08 Time:  0:00:08


...done, 499414571 bytes.
22


100% (583749490 of 583749490) |##########| Elapsed Time: 0:00:14 Time:  0:00:14


...done, 583749490 bytes.
23


100% (577213434 of 577213434) |##########| Elapsed Time: 0:00:31 Time:  0:00:31


...done, 577213434 bytes.
24


100% (524960256 of 524960256) |##########| Elapsed Time: 0:00:08 Time:  0:00:08


...done, 524960256 bytes.
25


100% (379068027 of 379068027) |##########| Elapsed Time: 0:00:08 Time:  0:00:08


...done, 379068027 bytes.
26


100% (557141269 of 557141269) |##########| Elapsed Time: 0:00:07 Time:  0:00:07


...done, 557141269 bytes.
27


100% (541793959 of 541793959) |##########| Elapsed Time: 0:00:15 Time:  0:00:15


...done, 541793959 bytes.
28


100% (577778044 of 577778044) |##########| Elapsed Time: 0:00:07 Time:  0:00:07


...done, 577778044 bytes.
29


100% (564636748 of 564636748) |##########| Elapsed Time: 0:00:14 Time:  0:00:14


...done, 564636748 bytes.
30


100% (549473815 of 549473815) |##########| Elapsed Time: 0:00:19 Time:  0:00:19


...done, 549473815 bytes.
31


100% (487371402 of 487371402) |##########| Elapsed Time: 0:00:14 Time:  0:00:14


...done, 487371402 bytes.
32


100% (580172291 of 580172291) |##########| Elapsed Time: 0:00:11 Time:  0:00:11


...done, 580172291 bytes.
33


100% (482606591 of 482606591) |##########| Elapsed Time: 0:00:08 Time:  0:00:08


...done, 482606591 bytes.
34


100% (556529331 of 556529331) |##########| Elapsed Time: 0:00:09 Time:  0:00:09


...done, 556529331 bytes.
35


100% (412442249 of 412442249) |##########| Elapsed Time: 0:00:07 Time:  0:00:07


...done, 412442249 bytes.
36


100% (546167222 of 546167222) |##########| Elapsed Time: 0:00:11 Time:  0:00:11


...done, 546167222 bytes.
37


100% (553049136 of 553049136) |##########| Elapsed Time: 0:00:18 Time:  0:00:18


...done, 553049136 bytes.
38


100% (570197407 of 570197407) |##########| Elapsed Time: 0:00:07 Time:  0:00:07


...done, 570197407 bytes.
39


100% (420518306 of 420518306) |##########| Elapsed Time: 0:00:07 Time:  0:00:07


...done, 420518306 bytes.
40


100% (483584042 of 483584042) |##########| Elapsed Time: 0:00:10 Time:  0:00:10


...done, 483584042 bytes.
41


100% (513259131 of 513259131) |##########| Elapsed Time: 0:00:08 Time:  0:00:08


...done, 513259131 bytes.
42


100% (561493142 of 561493142) |##########| Elapsed Time: 0:00:13 Time:  0:00:13


...done, 561493142 bytes.
43


100% (433012913 of 433012913) |##########| Elapsed Time: 0:00:09 Time:  0:00:09


...done, 433012913 bytes.
44


100% (561921244 of 561921244) |##########| Elapsed Time: 0:00:08 Time:  0:00:08


...done, 561921244 bytes.
45


HTTPError: HTTP Error 404: The specified blob does not exist.

## Code to Fix arrays 

Fix arrays when tiles are not properly allocated

In [ ]:
N =   #specify the number of tiles that were not properly allocated
#load the array
remaining = np.load('tile_name_tile_url_remaining_expanded.npy')
labeled = np.load("tile_name_tile_url_labeled.npy")
#Save the ammended arrays
np.save('tile_name_tile_url_labeled',labeled[:-N])
np.save('tile_name_tile_url_remaining_expanded',np.concatenate((labeled[-N:], remaining)))

Fix arrays when a tile cannot be found (HTTP Error 404: The specified blob does not exist.)

In [7]:
row_to_remove = 45

#dist.tile_name_tile_url_tiles_for_annotators[row_to_remove]

error_downloading = np.concatenate((np.load('tile_name_tile_url_error_downloading.npy'),
                                    dist.tile_name_tile_url_tiles_for_annotators[row_to_remove]))
np.save('tile_name_tile_url_error_downloading.npy', error_downloading)

#Remove the rows with errors
print(np.load('tile_name_tile_url_remaining_expanded.npy').shape)
np.save('tile_name_tile_url_remaining_expanded',np.delete(np.load(tiles_remaining), row_to_remove, axis = 0))
print(np.load('tile_name_tile_url_remaining_expanded.npy').shape)
row_to_remove = np.nan

(545, 2)
(544, 2)


In [ ]:
#recover the unlabeled images from the dist
## np.save('tile_name_tile_url_remaining_expanded',dist.tile_name_tile_url_unlabeled)
#dist.tile_name_tile_url_tiles_for_annotators

In [ ]:
#dist.track_tile_annotations(tiles_labeled)
#np.save('tile_name_tile_url_remaining_expanded', dist.tile_name_tile_url_remaining)
#np.save('tile_name_tile_url_labeled', dist.tile_name_tile_url_labeled)

## FIX Crop Script

In [ ]:
from PIL import Image

def crop(path, input, height, width, k, page, area):
    im = Image.open(input)
    imgwidth, imgheight = im.size
    for i in range(0,imgheight,height):
        for j in range(0,imgwidth,width):
            box = (j, i, j+width, i+height)
            a = im.crop(box)
            try:
                o = a.crop(area)
                o.save(os.path.join(path,"PNG","%s" % page,"IMG-%s.png" % k))
            except:
                pass
            k +=1

In [ ]:
dist.tile_names = os.listdir(dist.tiles_dir) #get a list of all of the tiles in tiles directory
for tile_name in dist.tile_names: #index over the tiles in the tiles_dir 
    file_name, ext = os.path.splitext(tile_name) # File name
    print(tile_name)

    print(count)

### Pil

In [ ]:
from PIL import Image 
 
#img = Image.open(r"C:\Users\rapiduser\Downloads\shutterstock_95235265.jpg") 
img = Image.open(r"C:\chips_allocation\Check\tiles\ca_060cm_2018_38121_m_3812128_ne_10_060_20180714.tif")
img.save('test.jpeg')
img = Image.open(r"C:\chips_allocation\Check\tiles\ca_060cm_2018_38121_m_3812128_ne_10_060_20180714.")

left = 0
top = 50
right = 510
bottom = 292
 
img_res = img.crop((left, top, right, bottom)) 
 
img_res.show() 

### Open CV

In [14]:
import cv2
import os
image = cv2.imread(r"C:/chips_allocation/Check/tiles/ca_060cm_2018_38121_m_3812128_ne_10_060_20180714.tif") 
y=0
x=512*19
h=512
w=512
#cropped = image[x:w, y:h]
cropped = image[y:y+h, x:x+w]
print(image.shape)

# Filename
filename = 'test2.jpg'
cv2.imwrite(os.path.join('C:/chips_allocation/Check/tiles',filename),cropped)
cv2.imshow("Cropped", cropped)
cv2.waitKey(0)
# Using cv2.imwrite() method
# Saving the image
#cv2.imwrite(filename, img)


(12220, 9770, 3)


-1

In [ ]:
from PIL import Image
Image.MAX_IMAGE_PIXELS = 999999999999
import os
file_name = "ca_060cm_2018_38121_m_3812128_ne_10_060_20180714"
#tile = Image.open(os.path.join(dist.tiles_dir, tile_name)) #open image
with Image.open(r"C:\chips_allocation\Check\tiles\ca_060cm_2018_38121_m_3812128_ne_10_060_20180714") as tile:
    width, height = tile.size
    item_width = int(512)
    col = int(width/512)+1
    row = int(height/512)+1
    count = 1
    for j in range(0,col):
        for i in range(0,row):
            box = (i*item_width, j*item_width, (i+1)*item_width, (j+1)*item_width)
            chips_save_path = os.path.join(self.chips_dir, file_name+ '_'+ \
                                           str(count).zfill(6) + '.jpg') 
            # The index is a six-digit number like '000023'. 
            #chip = tile.crop(box)
            #chip = chip.save(chips_save_path)
            chip = tile.crop(box).save(chips_save_path)
            count += 1
    print(count)

In [4]:
import numpy as np
import cv2
import os

image = cv2.imread(r"C:/chips_allocation/Check/tiles/ca_060cm_2018_38121_m_3812128_ne_10_060_20180714.tif") 
y=0
x=512*19
h=512
w=512
#cropped = image[x:w, y:h]
cropped = image[y:y+h, x:x+w]

print(image.shape)

# Filename
filename = 'test2.jpg'
cv2.imwrite(os.path.join('C:/chips_allocation/Check/tiles',filename),cropped)
cv2.imshow("Cropped", cropped)
cv2.waitKey(0)
# Using cv2.imwrite() method
# Saving the image
#cv2.imwrite(filename, img)

#image = cv2.imread('C:/chips_allocation/Check/tiles/test2.jpg') 
print(chip_img.width)
print(chip_img.height)

#At the end of the image
if chip_img.width != 512:
    black_width = 512 - chip_img.width
    black_height = 512
    black_image = np.zeros((black_height,black_width,3), np.uint8)
    print("Incorrect Width")
    print(black_img.width)
    print(black_img.height)
    chip_img = np.concatenate([chip_img, blank_image])
    
if chip_img.height != 512:
    black_height = 512 - chip_img.height
    black_width = 512
    black_image = np.zeros((black_height,black_width,3), np.uint8)
    print("Incorrect Height")
    print(black_img.width)
    print(black_img.height)
    chip_img = np.concatenate([chip_img, blank_image])

In [8]:
#blank_image = np.zeros((512,512-42,3), np.uint8)
cv2.imshow('3 Channel Window', 1)
print("image shape: ", blank_image.shape)
cv2.waitKey(0)
cv2.destroyAllWindows()

image shape:  (512, 470, 3)


In [35]:
from PIL import Image


#blackBox = Image.new(im.mode, (im.width, 50), '#000')
#im.paste(blackBox, (0, im.height - blackBox.height))
#for x in range(img.width):
#    for y in range(img.height - 40, img.height):
#        img.putpixel((x, y), (0, 0, 0))

#img.save('test2_bb.png')

42
512


In [40]:
from PIL import Image  
chip_img = Image.open('C:/chips_allocation/Check/tiles/test2.jpg')
print(chip_img.width)
print(chip_img.height)

if chip_img.width != 512:
    black_width = 512 - chip_img.width
    black_height = 512
    black_img = Image.new(mode = "RGB", size = (black_width, black_height))
    print("Incorrect Width")
    print(black_img.width)
    print(black_img.height)
    chip_img+black_img
    
if chip_img.height != 512:
    black_height = 512 - chip_img.height
    black_width = 512
    blackimg = Image.new(mode = "RGB", size = (black_width, black_height))
    print("Incorrect Height")
    print(black_img.width)
    print(black_img.height)

42
512
Incorrect Width
470
512


TypeError: unsupported operand type(s) for +: 'JpegImageFile' and 'Image'

In [ ]:
import numpy as np
import cv2

image = cv2.imread('download.jpg')
y=0
x=0
h=100
w=200
crop = image[y:y+h, x:x+w]
cv2.imshow('Image', crop)
cv2.waitKey(0) 

### Wand

In [7]:
from __future__ import print_function
from wand.image import Image

with Image(filename=r"C:/chips_allocation/Check/tiles/ca_060cm_2018_38121_m_3812128_ne_10_060_20180714.tif") as img:
    print('format =', img.format)
    print('width =', img.width)
    print('height =', img.height)

format = TIFF
width = 306
height = 382
